tests/test_edgeql_data_migration.py

```diff
@@ -0,0 +1,1345 @@
+
+        await self._migrate(r"""
+            type Child;
+
+            type Base {
+                required link bar -> Child {
+                    # add a constraint
+                    constraint exclusive;
+                }
+            }
+        """)
+
+        await self.assert_query_result(
+            r"""
+                SELECT Base {
+                    bar: {id},
+                };
+            """,
+            [{
+                'bar': {'id': res[0].bar.id},
+            }],
+        )
+
+    async def test_edgeql_migration_08(self):
+        await self.con.execute("""
+            SET MODULE test;
+        """)
+        await self._migrate(r"""
+            type Base {
+                property foo -> str;
+            }
+        """)
+        await self.con.execute(r"""
+            INSERT Base {
+                foo := 'very_long_test_str_base_08',
+            };
+        """)
+
+        # Try altering the schema to a state inconsistent with current
+        # data.
+        new_state = r"""
+            type Base {
+                required property foo -> str {
+                    # add a constraint
+                    constraint max_len_value(10);
+                }
+            }
+        """
+        with self.assertRaisesRegex(
+                edgedb.ConstraintViolationError,
+                r"foo must be no longer than 10 characters"):
+            await self._migrate(new_state)
+
+        # Fix the data.
+        await self.con.execute(r"""
+            UPDATE Base
+            SET {
+                foo := 'base_08',
+            };
+        """)
+
+        # Migrate to same state as before now that the data is fixed.
+        await self._migrate(new_state)
+
+        await self.assert_query_result(
+            r"""
+                SELECT Base {
+                    foo,
+                };
+            """,
+            [{
+                'foo': 'base_08',
+            }],
+        )
+
+    async def test_edgeql_migration_09(self):
+        await self.con.execute("""
+            SET MODULE test;
+        """)
+        await self._migrate(r"""
+            scalar type constraint_length extending str {
+                constraint max_len_value(10);
+            }
+            type Base {
+                property foo -> constraint_length;
+            }
+        """)
+        await self.con.execute(r"""
+            INSERT Base {
+                foo := 'b09',
+            };
+        """)
+
+        # Try altering the schema to a state inconsistent with current
+        # data.
+        new_state = r"""
+            scalar type constraint_length extending str {
+                constraint max_len_value(10);
+                # add a constraint
+                constraint min_len_value(5);
+            }
+            type Base {
+                property foo -> constraint_length;
+            }
+        """
+        with self.assertRaisesRegex(
+                edgedb.InternalServerError,
```

Comment:
Postgres errors should not leak, please don't write tests against them.  The error should be translated properly.